File: src/main/java/com/langleon/dsobuildsim/items/dto/ItemDefinitionDTO.java
package com.langleon.dsobuildsim.items.dto;

import com.langleon.dsobuildsim.common.StatType;
import com.langleon.dsobuildsim.enchantments.dto.EnchantmentDTO;
import com.langleon.dsobuildsim.items.core.enums.ItemType;
import com.langleon.dsobuildsim.items.core.enums.ItemCategory;
import com.langleon.dsobuildsim.items.core.enums.ItemSlotType;
import com.langleon.dsobuildsim.sets.SetType;

import java.util.Map;
import java.util.Set;

public record ItemDefinitionDTO(
        ItemCategory itemCategory,
        ItemType itemType,
        String name,
        int defaultLevel,
        int tier,
        ItemSlotType itemSlotType,
        Map<StatType, Double> rawBaseValues,

        //optional fields depending on category
        Map<StatType, Double> uniqueBaseValues,
        Map<StatType, Double> uniqueRelativeValues,
        Map<StatType, Double> uniqueAbsoluteValues,
        Set<EnchantmentDTO> uniqueEnchantments,
        String uniqueDescription,

        SetType set
        ) {
}

File: frontend/src/app/models/item-definition-dto.ts
import { EnchantmentDTO } from "./enchantment-dto";
import { StatType } from "./stat-type";

export interface ItemDefinitionDTO {
    itemCategory: string;
    itemType: string;
    name: string;
    defaultLevel: number;
    tier: number;
    itemSlotType: string;
    rawBaseValues: Record<StatType, number>;
    //optional fields, might be empty
    uniqueBaseValues: Record<StatType, number>;
    uniqueRelativeValues: Record<StatType, number>;
    uniqueAbsoluteValues: Record<StatType, number>;
    uniqueEnchantments: EnchantmentDTO[];
    uniqueDescription: string;
    set: string;
}